--- src/scanner.c
+++ src/scanner.c
@@ -291,11 +291,48 @@ static bool scan_end_delimiter(Scanner *scanner, TSLexer *lexer) {
 }

 static bool scan_comment_content(Scanner *scanner, TSLexer *lexer) {
-  int first = get_delimiter(scanner->end_delimiter, 0, DEFAULT_END_DELIMITER);
-  while (lexer->lookahead != first) {
-    if (lexer->eof(lexer))
+  lexer->mark_end(lexer);
+  int end_delimiter_max = scanner->end_delimiter.size == 0
+                              ? DEFAULT_SIZE
+                              : scanner->end_delimiter.size;
+  int current_size = 0;
+  int end_i = 0;
+
+  while (true) {
+    int ith_end =
+        get_delimiter(scanner->end_delimiter, end_i, DEFAULT_END_DELIMITER);
+
+    if (lexer->lookahead == ith_end) {
+      end_i++;
+      lexer->advance(lexer, false);
+    } else {
+      if (end_i > 0) {
+        for (int i = 0; i < end_i; i++) {
+          lexer->mark_end(lexer);
+          current_size++;
+        }
+      }
+      lexer->advance(lexer, false);
+      lexer->mark_end(lexer);
+      current_size++;
+      end_i = 0;
+    }
+
+    if (end_i == end_delimiter_max && current_size > 0)
+      break;
+    else if (end_i == end_delimiter_max && current_size == 0)
+      return false;
+
+    if (lexer->eof(lexer) && current_size > 0) {
+      if (end_i > 0) {
+        for (int i = 0; i < end_i; i++) {
+          lexer->mark_end(lexer);
+          current_size++;
+        }
+      }
+      break;
+    } else if (lexer->eof(lexer) && current_size == 0)
       return false;
-    lexer->advance(lexer, false);
   }
   lexer->result_symbol = COMMENT_CONTENT;
   return true;
